src/oscore/oscore_cbor.c

```diff
@@ -0,0 +1,430 @@
+  if (value < 0x18) { /* small value half a byte */
+    assert(*buf_size >= 1);
+    (*buf_size)--;
+    (**buffer) = (uint8_t)value;
+    (*buffer)++;
+    return 1;
+  } else if ((value > 0x17) && (value < 0x100)) {
+    /* one byte uint8_t  */
+    assert(*buf_size >= 2);
+    (*buf_size) -= 2;
+    (**buffer) = (0x18);
+    *buffer = (*buffer) + 2;
+    put_b_f(buffer, value, 0);
+    return 2;
+  } else if ((value > 0xff) && (value < 0x10000)) {
+    /* 2 bytes uint16_t     */
+    assert(*buf_size >= 3);
+    (*buf_size) -= 3;
+    (**buffer) = (0x19);
+    *buffer = (*buffer) + 3;
+    put_b_f(buffer, value, 1);
+    return 3;
+  } else if ((value > 0xffff) && (value < 0x100000000)) {
+    /* 4 bytes uint32_t   */
+    assert(*buf_size >= 5);
+    (*buf_size) -= 5;
+    (**buffer) = (0x1a);
+    *buffer = (*buffer) + 5;
+    put_b_f(buffer, value, 3);
+    return 5;
+  } else /*if(value > 0xffffffff)*/ {
+    /* 8 bytes uint64_t  */
+    assert(*buf_size >= 9);
+    (*buf_size) -= 9;
+    (**buffer) = (0x1b);
+    *buffer = (*buffer) + 9;
+    put_b_f(buffer, value, 7);
+    return 9;
+  }
+}
+
+/* temporary routine to read ascii hex dump */
+static uint8_t
+gethex_byte(const uint8_t *buffer) {
+  /*
+     uint8_t temp1 = *buffer;
+     if ('0' <= temp1 && temp1 <= '9') temp1 = temp1 - '0';
+     else if ('A' <= temp1 && temp1 <= 'F')
+                                  temp1 = 10 + temp1 - 'A';
+     else if ('a' <= temp1 && temp1 <= 'f')
+                                  temp1 = 10 + temp1 - 'a';
+     else return 255;
+     uint8_t temp2 = *buffer;
+     if ('0' <= temp2 && temp2 <= '9') temp2 = temp2 - '0';
+     else if ('A' <= temp2 && temp2 <= 'F')
+                                  temp2 = 10 + temp2 - 'A';
+     else if ('a' <= temp2 && temp2 <= 'f')
+                                  temp2 = 10 + temp2 - 'a';
+     else return 255;
+     return (temp1<<4) + temp2;
+  */
+  return *buffer;
+}
+
+uint8_t
+oscore_cbor_get_next_element(const uint8_t **buffer) {
+  uint8_t element = gethex_byte(*buffer);
+  return element >> 5;
+}
+
+/* oscore_cbor_get_element_size returns
+ *   - size of byte strings of character strings
+ *   - size of array
+ *   - size of map
+ *   - value of unsigned integer
+ */
+
+uint64_t
+oscore_cbor_get_element_size(const uint8_t **buffer) {
+  uint8_t control = gethex_byte((*buffer)) & 0x1f;
+  uint64_t size = gethex_byte((*buffer)++);
+  if (control < 0x18) {
+    size = (uint64_t)control;
+  } else {
+    control = control & 0x3;
+    int num = 1 << control;
+    size = 0;
+    uint64_t getal;
+    for (int i = 0; i < num; i++) {
+      getal = (uint64_t)gethex_byte((*buffer)++);
+      size = (size << 8) + getal;
+    }
+  }
+  return size;
+}
+
+uint8_t
+oscore_cbor_elem_contained(const uint8_t *data, uint8_t *end) {
+  const uint8_t *buf = data;
+  const uint8_t *last = data + oscore_cbor_get_element_size(&buf);
+  if (last > end) {
+    fprintf(stderr, "oscore_cbor_elem_contained returns 1 \n");
```

Comment:
I think that `coap_log()` should be used here.

Reply:
Agreed.  Updated.